packages/core/src/new-api/internal/execution/execution-engine.ts

```diff
@@ -111,25 +124,39 @@ export class ExecutionEngine {
       })
     );
 
-    while (!isExecutionResult(current)) {
+    while (!isExecutionResultMessage(current)) {
       context.executionState = state.executionStateMap[future.id];
 
       if (isOnChainAction(current)) {
         current = await state.transactionService.onchain(current, {
           libraries,
         });
-      } else if (isOnchainResult(current)) {
+      } else if (isOnChainSuccessMessage(current)) {
         current = (await exectionStrategy.next(current)).value;
+      } else if (isOnChainFailureMessage(current)) {
+        current = this._convertToExecutionFailure(current);
       } else {
         current = (await exectionStrategy.next(null)).value;
       }
 
+      // TODO: add never type check?
```

Comment:
I guess there's no other way to do this, because a switch wouldn't work.

Reply:
I do think that we need some sort of check, because in the latest else `current` can have quiet a few types, some of them unrelated, like `WipeMessage`.

I guess the least we can do is avoid using plain `else` (without if) in these chains of `if`s

Reply:
I am going to add an assertNever case to typecheck that all message types are covered, in a later PR.